File: lib/domain/entities/local_db/post_entity.dart
// lib/domain/entities/local_db/post_entity.dart

/// Represents a post within the app's domain.
///
/// Includes various metadata: unique IDs, timestamps, images, location info, etc.
class PostEntity {
  /// A locally generated ID (UUID), used to identify the post in local DB.
  final String postIdLocal;

  /// The published ID from the remote platform (X/Twitter), if available.
  final String? postIdX;

  /// Main text content of the post.
  final String content;

  /// A short title or heading for the post.
  final String title;

  /// Timestamp when the post was created locally.
  final DateTime createdAt;

  /// Timestamp when the post was last updated locally, if any.
  final DateTime? updatedAt;

  /// Timestamp for when the post is scheduled to be published, if any.
  final DateTime? scheduledAt;

  /// Who can see or reply to the post. Typically 'everyone' or similar.
  final String? visibility;

  /// Local file paths for images included with this post.
  final List<String> localImagePaths;

  /// Remote URLs (e.g., Firebase Storage) for images, if already uploaded.
  final List<String> cloudImageUrls;

  /// Optional latitude for location-based posts.
  final double? locationLat;

  /// Optional longitude for location-based posts.
  final double? locationLng;

  /// Optional human-readable address or location text.
  final String? locationAddress;

  /// Constructs a [PostEntity] with required and optional fields.
  PostEntity({
    required this.postIdLocal,
    required this.postIdX,
    required this.content,
    required this.title,
    required this.createdAt,
    this.updatedAt,
    this.scheduledAt,
    this.visibility,
    required this.localImagePaths,
    required this.cloudImageUrls,
    this.locationLat,
    this.locationLng,
    this.locationAddress,
  });

  /// Returns a copy of this entity with any specified fields replaced.
  PostEntity copyWith({
    String? postIdLocal,
    String? postIdX,
    String? content,
    String? title,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? scheduledAt,
    String? visibility,
    List<String>? localImagePaths,
    List<String>? cloudImageUrls,
    double? locationLat,
    double? locationLng,
    String? locationAddress,
  }) {
    return PostEntity(
      postIdLocal: postIdLocal ?? this.postIdLocal,
      postIdX: postIdX ?? this.postIdX,
      content: content ?? this.content,
      title: title ?? this.title,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      scheduledAt: scheduledAt ?? this.scheduledAt,
      visibility: visibility ?? this.visibility,
      localImagePaths: localImagePaths ?? this.localImagePaths,
      cloudImageUrls: cloudImageUrls ?? this.cloudImageUrls,
      locationLat: locationLat ?? this.locationLat,
      locationLng: locationLng ?? this.locationLng,
      locationAddress: locationAddress ?? this.locationAddress,
    );
  }
}

File: vouse_server/src/posts/entities/post.entity.ts
// src/posts/entities/post.entity.ts
import {
  Entity,
  Column,
  PrimaryGeneratedColumn,
  CreateDateColumn,
  UpdateDateColumn,
  ManyToOne,
  JoinColumn,
} from 'typeorm';
import { User } from '../../users/entities/user.entity';
import { ApiProperty, ApiPropertyOptional } from '@nestjs/swagger';

// Post status enum
export enum PostStatus {
  DRAFT = 'draft',
  SCHEDULED = 'scheduled',
  PUBLISHED = 'published',
  FAILED = 'failed',
}

/**
 * Post entity for storing social media posts
 * Includes scheduling, content, and metadata
 */
@Entity('posts')
export class Post {
  /**
   * Primary key UUID
   */
  @ApiProperty({
    description: 'Primary key UUID for the post',
    example: 'a1b2c3d4-e5f6-7890-1234-567890abcdef',
  })
  @PrimaryGeneratedColumn('uuid')
  id: string;

  /**
   * Client-generated unique ID for the post
   * Used for tracking the post across client and server
   */
  @ApiProperty({
    description: 'Client-generated unique ID for tracking',
    example: 'local-uuid-123',
  })
  @Column({ name: 'post_id_local', type: 'text', unique: true })
  postIdLocal: string;

  /**
   * X/Twitter post ID for published posts
   */
  @ApiPropertyOptional({
    description: 'X/Twitter post ID after publication',
    example: '1773600000000000000',
  })
  @Column({ name: 'post_id_x', type: 'text', nullable: true })
  postIdX: string | null = null;

  /**
   * User ID who created the post
   */
  @ApiProperty({
    description: 'Firebase User ID of the post creator',
    example: 'firebaseUid123',
  })
  @Column({ name: 'user_id', type: 'text' })
  userId: string;

  /**
   * Relation to User entity
   */
  @ManyToOne(() => User, { eager: true })
  @JoinColumn({ name: 'user_id' })
  user: User;

  /**
   * Post content/message
   */
  @ApiProperty({
    description: 'The main text content of the post',
    example: 'Hello Twitter!',
  })
  @Column({ type: 'text' })
  content: string;

  /**
   * Optional post title
   */
  @ApiPropertyOptional({
    description: 'An optional internal title for the post',
    example: 'My First Vouse Post',
  })
  @Column({ type: 'text', nullable: true })
  title: string | null = null;

  /**
   * Scheduled publication date for the post
   */
  @ApiPropertyOptional({
    description:
      'ISO 8601 timestamp when the post is scheduled for publication',
    type: 'string',
    format: 'date-time',
    example: '2025-04-01T10:00:00Z',
  })
  @Column({ name: 'scheduled_at', type: 'timestamp', nullable: true })
  scheduledAt: Date | null = null;

  /**
   * Actual publication date for the post
   */
  @ApiPropertyOptional({
    description: 'ISO 8601 timestamp when the post was actually published',
    type: 'string',
    format: 'date-time',
    example: '2025-04-01T10:00:05Z',
  })
  @Column({ name: 'published_at', type: 'timestamp', nullable: true })
  publishedAt: Date | null = null;

  /**
   * Current status of the post
   * draft, scheduled, published, or failed
   */
  @ApiProperty({
    description: 'Current status of the post',
    enum: PostStatus,
    example: PostStatus.SCHEDULED,
  })
  @Column({
    type: 'enum',
    enum: PostStatus,
    default: PostStatus.DRAFT,
  })
  status: PostStatus = PostStatus.DRAFT;

  /**
   * Reason for failure if status is 'failed'
   */
  @ApiPropertyOptional({
    description: 'Reason for publication failure, if applicable',
    example: 'Twitter API error: Invalid credentials.',
  })
  @Column({ name: 'failure_reason', type: 'text', nullable: true })
  failureReason: string | null = null;

  /**
   * Visibility setting (public, private, etc.)
   */
  @ApiPropertyOptional({
    description: 'Visibility setting (currently unused)',
    example: 'public',
  })
  @Column({ type: 'text', nullable: true })
  visibility: string | null = null;

  /**
   * URLs for images to include with the post
   */
  @ApiPropertyOptional({
    description: 'Array of image URLs from cloud storage',
    example: ['https://storage.googleapis.com/.../image1.jpg'],
    type: [String],
  })
  @Column('simple-array', { name: 'cloud_image_urls', default: [] })
  cloudImageUrls: string[] = [];

  /**
   * Optional location latitude
   */
  @ApiPropertyOptional({
    description: 'Latitude for the post location',
    example: 34.0522,
  })
  @Column({ name: 'location_lat', type: 'float', nullable: true })
  locationLat: number | null = null;

  /**
   * Optional location longitude
   */
  @ApiPropertyOptional({
    description: 'Longitude for the post location',
    example: -118.2437,
  })
  @Column({ name: 'location_lng', type: 'float', nullable: true })
  locationLng: number | null = null;

  /**
   * Optional location address or name
   */
  @ApiPropertyOptional({
    description: 'Address or name of the location',
    example: 'Los Angeles, CA',
  })
  @Column({ name: 'location_address', type: 'text', nullable: true })
  locationAddress: string | null = null;

  /**
   * Creation timestamp
   */
  @ApiProperty({
    description: 'Timestamp when the post record was created',
    type: 'string',
    format: 'date-time',
  })
  @CreateDateColumn({ name: 'created_at' })
  createdAt: Date = new Date();

  /**
   * Last update timestamp
   */
  @ApiProperty({
    description: 'Timestamp when the post record was last updated',
    type: 'string',
    format: 'date-time',
  })
  @UpdateDateColumn({ name: 'updated_at' })
  updatedAt: Date = new Date();
}
